src/jwt-utils/jwtdecode.ts

```diff
@@ -0,0 +1,192 @@
+
+  let issuerOptions: Record<string, unknown> = {}
+  if (typeof pubkeyOrSharedKey === 'object' && pubkeyOrSharedKey !== null && pubkeyOrSharedKey['publicKey']) {
+    issuerOptions = pubkeyOrSharedKey
+    pubkeyOrSharedKey = pubkeyOrSharedKey['publicKey']
+  }
+
+  if (!pubkeyOrSharedKey) {
+    throw new JwtVerifyError(`Unknown pubkey id '${header.kid}' for this issuer`)
+  }
+
+  const signatureOrHash = base64UrlSafe.decode(parts[2])
+  /* istanbul ignore else */
+  if (signAlgo) {
+    // Validate signature
+    const verifier = crypto.createVerify(signAlgo)
+    verifier.write(`${parts[0]}.${parts[1]}`, 'utf8')
+    verifier.end()
+    if (!verifier.verify(pubkeyOrSharedKey, signatureOrHash)) {
+      throw new JwtVerifyError(`Signature verification failed with alg '${header.alg}'`)
+    }
+  } else if (hmacAlgo) {
+    const hmac = crypto.createHmac(hmacAlgo, pubkeyOrSharedKey)
+    hmac.update(`${parts[0]}.${parts[1]}`, 'utf8')
+    const signatureBuffer = hmac.digest()
+    if (!crypto.timingSafeEqual(signatureOrHash, signatureBuffer)) {
+      throw new JwtVerifyError(`Verification failed with alg '${header.alg}'`)
+    }
+  } else {
+    throw Error(`Should never happen`)
+  }
+
+  const unixNow = Math.floor(Date.now() / 1000)
+  const validators = {
+    aud: validateAudience,
+    exp: validateExpires,
+    iat: validateIssuedAt,
+    nbf: validateNotBefore
+  }
+  Object.assign(validators, options.validators || {})
+  Object.assign(validators, issuerOptions.validators || {})
+
+  const validationOptions = {}
+  Object.assign(validationOptions, options)
+  Object.assign(validationOptions, issuerOptions)
+
+  validators.aud(body, audiences)
+  validators.iat(body, unixNow, validationOptions)
+  validators.nbf(body, unixNow, validationOptions)
+  validators.exp(body, unixNow, validationOptions)
+
+  return body
+}
+
+function validateNotBefore(body: Record<string, unknown>, unixNow: number, options: Record<string, unknown>): void {
+  if (typeof options.nbfIatSkew === 'number') {
+    if (body.nbf && body.nbf > unixNow + options.nbfIatSkew) {
+      throw new JwtVerifyError(`Not before in the future by more than ${options.nbfIatSkew} seconds`)
+    }
+  }
+}
+
+function validateIssuedAt(body: Record<string, unknown>, unixNow: number, options: Record<string, unknown>): void {
+  if (typeof options.nbfIatSkew === 'number') {
+    if (body.iat && body.iat > unixNow + options.nbfIatSkew) {
+      throw new JwtVerifyError(`Issued at in the future by more than ${options.nbfIatSkew} seconds`)
+    }
+  }
+}
+
+function validateAudience(body: Record<string, unknown>, audiences: string[]): void {
+  const auds = Array.isArray(body.aud) ? body.aud : [body.aud]
+  if (!auds.some(aud => audiences.includes(aud))) {
+    throw new JwtVerifyError(`Unknown audience '${auds.join(',')}'`)
+  }
+}
+
+function validateExpires(body: Record<string, unknown>, unixNow: number, options: Record<string, unknown>): void {
```

Comment:
Do we have a JWT type somewhere that we could use instead of `Record<string, unknown>`?

Reply:
Then you could also remove a lot of the type checks that you currently need, because you can't infer the type